### PR TITLE
Adds grpc to the scheduler

### PR DESCRIPTION
**Pull request by jcooklin**
* This PR is part of supporting distributed workflows

### CODE REVIEW

File: mgmt/rest/task.go [hunk 1]
```diff
@@ -75,51 +80,56 @@ func (s *Server) addTask(w http.ResponseWriter, r *http.Request, _ httprouter.Pa
 		respond(500, rbody.FromError(err), w)
 		return
 	}
-
-	sch, err := makeSchedule(tr.Schedule)
+	sch, err := json.Marshal(tr.Schedule)
 	if err != nil {
-		respond(500, rbody.FromError(err), w)
+		respond(404, rbody.FromError(err), w)
```
Comment (danielscottt) [inline mgmt/rest/task.go]: `404`?  Maybe `400` (bad request)?

File: mgmt/rest/task.go [hunk 2]
```diff
@@ -75,51 +80,56 @@ func (s *Server) addTask(w http.ResponseWriter, r *http.Request, _ httprouter.Pa
...
-	if tr.Deadline != "" {
-		dl, err := time.ParseDuration(tr.Deadline)
-		if err != nil {
-			respond(500, rbody.FromError(err), w)
-			return
-		}
-		opts = append(opts, core.TaskDeadlineDuration(dl))
+	wm, err := json.Marshal(tr.Workflow)
+	if err != nil {
+		respond(404, rbody.FromError(err), w)
```
Comment (danielscottt) [inline mgmt/rest/task.go]: Same here.

File: mgmt/rest/task.go [hunk 3]
```diff
@@ -75,51 +80,56 @@ func (s *Server) addTask(w http.ResponseWriter, r *http.Request, _ httprouter.Pa
...
+			TaskName: tr.Name,
+		},
 	}
-	opts = append(opts, core.OptionStopOnFailure(10))
-
-	task, errs := s.mt.CreateTask(sch, tr.Workflow, tr.Start, opts...)
-	if errs != nil && len(errs.Errors()) != 0 {
+	t, err := s.mt.CreateTask(context.Background(), arg)
+	if err != nil {
+		respond(404, rbody.FromError(err), w)
```
Comment (danielscottt) [inline mgmt/rest/task.go]: `500` here probably.

File: mgmt/rest/task.go [hunk 4]
```diff
@@ -75,51 +80,56 @@ func (s *Server) addTask(w http.ResponseWriter, r *http.Request, _ httprouter.Pa
...
+	taskB := rbody.AddSchedulerTaskFromTask(t.Task)
+	taskB.Href = taskURI(r.Host, t.Task)
 	respond(201, taskB, w)
 }
 
 func (s *Server) getTasks(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
-	sts := s.mt.GetTasks()
+	reply, err := s.mt.GetTasks(context.Background(), &rpc.Empty{})
+	if err != nil {
+		respond(404, rbody.FromError(err), w)
```
Comment (danielscottt) [inline mgmt/rest/task.go]: `500`.

File: mgmt/rest/task.go [hunk 5]
```diff
@@ -130,9 +140,9 @@ func (s *Server) getTasks(w http.ResponseWriter, r *http.Request, _ httprouter.P
 
 func (s *Server) getTask(w http.ResponseWriter, r *http.Request, p httprouter.Params) {
 	id := p.ByName("id")
-	t, err1 := s.mt.GetTask(id)
-	if err1 != nil {
-		respond(404, rbody.FromError(err1), w)
+	t, err := s.mt.GetTask(context.Background(), &rpc.GetTaskArg{Id: id})
+	if err != nil {
+		respond(404, rbody.FromError(err), w)
```
Comment (danielscottt) [inline mgmt/rest/task.go]: `500`

Comment (danielscottt) [inline mgmt/rest/task.go]: Actually, maybe 404 makes sense here, I assume it's "task not found?"

File: mgmt/rest/task.go [hunk 6]
```diff
@@ -190,62 +193,63 @@ func (s *Server) watchTask(w http.ResponseWriter, r *http.Request, p httprouter.
...
+
+	if reply.Errors != nil {
+		if strings.Contains(reply.Errors[0].ErrorString, ErrTaskNotFound.Error()) {
+			respond(404, rbody.FromSnapErrors(rpc.ConvertSnapErrors(reply.Errors)), w)
 			return
 		}
-		if strings.Contains(errs[0].Error(), ErrTaskDisabledNotRunnable.Error()) {
-			respond(409, rbody.FromSnapErrors(errs), w)
+		if strings.Contains(reply.Errors[0].ErrorString, ErrTaskDisabledNotRunnable.Error()) {
+			respond(404, rbody.FromSnapErrors(rpc.ConvertSnapErrors(reply.Errors)), w)
```
Comment (danielscottt) [inline mgmt/rest/task.go]: Same here - Not 404.  Maybe `423 Locked`, or more simply, `403 Forbidden`?

Comment (ConnorDoyle) [inline mgmt/rest/task.go]: Another option is `409`:

> _409 Conflict_
>
> The request could not be completed due to a conflict with the current state of the resource. This code is only allowed in situations where it is expected that the user might be able to resolve the conflict and resubmit the request. The response body SHOULD include enough information for the user to recognize the source of the conflict. Ideally, the response entity would include enough information for the user or user agent to fix the problem; however, that might not be possible and is not required.

Source: https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html

Comment (danielscottt) [inline mgmt/rest/task.go]: no doubt.  `409` sounds best.

**Comment by geauxvirtual** [comment]
Speaking with @danielscottt yesterday, we want to land this in a feature branch first due to the number of changes in the PR and to ensure this is feature complete before landing in master. From reviewing the PR, I see three outstanding items that would need to be completed before this could land in master.

Support for all schedule types. This currently only looks to support simple schedule and landing this in master as is would remove a feature, windowed schedule, from snap.

Support for setting listening port and bind address for scheduler RPC server when snapd is starting.

Securing grpc communications between client and server.

**Comment by jcooklin** [comment]
PR moved to feature branch #748.

**Pull request closed without merging**